profiler/demangle.go

```diff
@@ -0,0 +1,39 @@
+package profiler
+
+import (
+	"bytes"
+	"compress/gzip"
+	"io"
+
+	"github.com/ianlancetaylor/demangle"
+	"gopkg.in/DataDog/dd-trace-go.v1/profiler/internal/pproflite"
+)
+
+// demangleCPUProfile demangles C++ or Rust symbol names appearing in the string
+// table for the given profile, and returns a new profile.
+func demangleCPUProfile(p []byte) ([]byte, error) {
+	r, err := gzip.NewReader(bytes.NewReader(p))
+	if err != nil {
+		return nil, err
+	}
+	p, err = io.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	out := new(bytes.Buffer)
+	w := gzip.NewWriter(out)
+	enc := pproflite.NewEncoder(w)
+
+	err = pproflite.NewDecoder(p).FieldEach(func(field pproflite.Field) error {
+		switch t := field.(type) {
+		case *pproflite.StringTable:
+			demangled := demangle.Filter(string(t.Value))
+			v := &pproflite.StringTable{Value: []byte(demangled)}
+			return enc.Encode(v)
```

Comment:
NITs:

1. This applies demangling logic to all strings. Any chance this will break stuff like file paths?
2. No need to allocate a new `pproflite.StringTable`, you can just modify `t.Value`. Yes, it's a little dirty, but ironically avoiding garbage is a dirty business 😆

Reply:
RE 1: So, the documentation for that function says it will return the input string unchanged if it's not a mangled symbol name. I haven't rigorously tested this, but I kind of assumed that would be okay for any inputs and we wouldn't need to be too rigorous about checking whether the input is even a symbol name. We could keep track the string table indices used for function names, though, and only check those.

RE 2, and optimizing this in general: The API from github.com/ianlancetaylor/demangle only works with strings, so we're probably paying at least one allocation for every demangling. I added a quick check on the prefix of each string to avoid things that don't have known mangled symbol prefixes. This lets us skip some byte-to-string conversions. But for super performance we'd probably want an append-based API, like `func AppendDemangeld(dst, input []byte) ([]byte, error)`. That way we can reuse a byte slice.